### PR TITLE
Refactor int4 and int8 weight only quantization to use `quantize`

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
This is similar to https://github.com/pytorch/ao/pull/294 but applied for int4 weight only quantization

Test Plan:

unit perf test:
python test/quantization/test_quant_api.py -k test_quantized_tensor_subclass_int4_wo_quant_perf
elapsed time: 0.2166275215148926, ref elapsed time: 0.2191881561279297
elapsed time: 0.2376406478881836, ref elapsed time: 0.22721023559570314
elapsed time: 0.21919679641723633, ref elapsed time: 0.2154969596862793

integration perf test:

reference: elapsed_time:  2.5900126953125  milliseconds
after refactor: elapsed_time:  2.56680078125  milliseconds
diff: no diff

TORCH_LOGS='output_code' python tutorials/quantize_vit/run_vit_b_quant.py

Before:
After:
generated code diff:

Reviewers:

Subscribers:

Tasks:

Tags:

Refactor int8 weight only quant to use quantize #299 logs

Summary:
Similar to https://github.com/pytorch/ao/pull/294 we replaced the implementation
of int8 weight only quant to used the newly added quantize function, as a part of
the unification effort for affine quantization

Test Plan:

unit perf test:
python test/quantization/test_quant_api.py -k test_quantized_tensor_subclass_int8_wo_quant_perf
elapsed time: 0.23909856796264647, ref elapsed time: 0.25150911331176756
elapsed time: 0.24894208908081056, ref elapsed time: 0.2570047950744629
elapsed time: 0.21607391357421876, ref elapsed time: 0.22809568405151368

integration test:
TORCH_LOGS='output_code' python tutorials/quantize_vit/run_vit_b_quant.py

Reference: elapsed_time: 1.355208740234375 milliseconds
After refactor: elapsed_time: 1.32778857421875 milliseconds

code diff (gist): [gist.github.com/jerryzh168/921a722cf20d476c8fc5888482e722dc](https://gist.github.com/jerryzh168/921a722cf20d476c8fc5888482e722dc)
code diff (meta-only paste): [internalfb.com/phabricator/paste/view/P1387333845](https://www.internalfb.com/phabricator/paste/view/P1387333845)

### CODE REVIEW

**Comment by jerryzh168** [comment]
this is rebased on int8-wo PR (https://github.com/pytorch/ao/pull/299) so will need to update this PR after the int8-wo PR is landed

File: torchao/dtypes/aqt.py
Comment (msaroufim) [inline torchao/dtypes/aqt.py:641]: n00b q: why does this require a transpose?

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: test/integration/test_integration.py
Comment (jerryzh168) [reply test/integration/test_integration.py:933]: I'm planning to open a PR after this, I have opened one for the other skip here: https://github.com/pytorch/ao/pull/300

File: torchao/dtypes/aqt.py
Comment (jerryzh168) [reply torchao/dtypes/aqt.py:81]: this means constructor, since we are returning `class.from_plain` now

File: torchao/dtypes/aqt.py
Comment (jerryzh168) [reply torchao/dtypes/aqt.py:267]: this one needs a bit more discussions with pt core team

File: torchao/dtypes/aqt.py
Comment (jerryzh168) [reply torchao/dtypes/aqt.py:295]: yeah, I'm confirming with @HDCharles right now, I think this is pretty weird, see comments in L575 of aqt.py for more details

File: torchao/dtypes/aqt.py
Comment (jerryzh168) [inline torchao/dtypes/aqt.py:575]: @msaroufim see this comment for more details

File: torchao/dtypes/aqt.py
Comment (jerryzh168) [reply torchao/dtypes/aqt.py:473]: this is specific to tinygemm kernels I think, copied from old code: https://github.com/pytorch/ao/blob/8a4e6938dbf52011483a0a22deb7254d4e9a29bc/torchao/quantization/subclass.py#L585-L586

File: torchao/dtypes/aqt.py
Comment (jerryzh168) [reply torchao/dtypes/aqt.py:641]: this is to align the dimensions, for block_size so that we can get groupsize from block_size argument, see L662, and also related to L575. right now the `_quantized_linear` does not have a well-defined accepted weight dimensions

File: tutorials/quantize_vit/run_vit_b_quant.py
Comment (jerryzh168) [reply tutorials/quantize_vit/run_vit_b_quant.py:17]: sure, this is just for people to easily try out different APIs, but we can just ask people to copy paste from README as well

File: torchao/dtypes/aqt.py
Comment (msaroufim) [reply torchao/dtypes/aqt.py:82]: This needs a comment I don't believe ctr is a common abbreviation for constructor

**Review by msaroufim — APPROVED** [review]

File: torchao/quantization/quant_api.py
Comment (cpuhrsch) [inline torchao/quantization/quant_api.py:232]: blind kwargs make it impossible to document the behavior. i understand that `change_linear_weights_to_int4_woqtensors` has this as well. Seems like something that could be worth fixing.

File: torchao/utils.py
Comment (cpuhrsch) [inline torchao/utils.py:63]: why?

File: torchao/quantization/quant_api.py
Comment (jerryzh168) [reply torchao/quantization/quant_api.py:232]: yeah sure

File: torchao/utils.py
Comment (jerryzh168) [reply torchao/utils.py:63]: we now use this in torchao/dtypes and torchao/quantization and have to do import tricks to avoid circular dep

**Comment by cpuhrsch** [comment]
Please don't merge PRs when CI is red and we can't get signal for incremental changes. Fix main CI first, then merge.

**Pull request merged**
